src/engine/Engine.cpp

```diff
@@ -1920,23 +1920,68 @@ void Engine::updateDirections()
 
 void Engine::updateScores()
 {
-    _candidatePlConstraints.clear();
-    for ( const auto plConstraint : _plConstraints )
+    if ( _networkLevelReasoner && GlobalConfiguration::SPLITTING_HEURISTICS ==
+         DivideStrategy::Polarity )
+    {
+        // We find the earliest K ReLUs that have not been fixed, update
+        // their scores, and pop them to the _candidatePlConstraints
+        // K is equal to GlobalConfiguration::RUNTIME_ESTIMATE_THRESHOLD
+        log( Stringf( "Using polarity heuristics..." ) );
+        for ( unsigned layer = 1; layer < _networkLevelReasoner->
+                  getNumberOfLayers() - 1; ++layer )
+        {
+            log( Stringf( "examining layer %u", layer ) );
+            unsigned size = _networkLevelReasoner->getLayerSize( layer );
+            for ( unsigned neuron = 0; neuron < size; ++neuron ){
+                auto plConstraint = _networkLevelReasoner->
+                    getPLConstraintFromIndex( layer, neuron );
+                if ( plConstraint && plConstraint->isActive()
+                     && !plConstraint->phaseFixed() )
+                {
+                    plConstraint->updateScore();
+                    _candidatePlConstraints.insert( plConstraint );
+                    if ( _candidatePlConstraints.size() >=
```

Comment:
I would suggest putting this check in the outer for loop and removing the return statements.

Reply:
Made the change accordingly.